presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -1121,6 +1110,15 @@ protected Type visitNode(Node node, StackableAstVisitorContext<Context> context)
             throw new SemanticException(NOT_SUPPORTED, node, "not yet implemented: " + node.getClass().getName());
         }
 
+        public Type visitGroupingOperation(GroupingOperation node, StackableAstVisitorContext<Context> context)
+        {
+            for (Expression columnArgument : node.getGroupingColumns()) {
+                process(columnArgument, context);
+            }
+            expressionTypes.put(node, INTEGER);
```

Comment:
So I ran into some problems converting the return type of `grouping` to BIGINT. When the query only has a `GROUP BY` and no `GROUPING SETS`/`ROLLUP`/`CUBE`, then the `grouping` should be re-written to a 0 literal. The problem is that the 0 literal is analysed in `ExpressionAnalyzer` and an INTEGER type is returned causing a mismatch and failing the query.

So I see two solutions: 1) leave the return type as INTEGER and lower the max number of columns that `grouping` can work on or 2) change the return type to BIGINT and in the case of just a `GROUP BY`, re-write the function with a special literal (e.g. -1) in the place of the `groupId` symbol and at runtime if you see that literal just output 0.

From my reading of the spec it seem that as long as we're using an exact numeric type (INTEGER or BIGINT) then we should be good. It doesn't say anything about varying the return type. See relevant portion quoted below SQL 2011 6.9.10:

> 10) If a \<grouping operation\> is specified, then:
>
> a) Let T be the aggregation query of \<set function specification\> that contains \<grouping operation\>. Each \<column reference\> shall reference a grouping column of T.
>
> b) The declared type of the result is exact numeric with an implementation-defined precision and a scale of 0 (zero).
>
> c) If more than one \<column reference\> is specified, then let N be the number of \<column reference\>s and let CRi, 1 (one) ≤ i ≤ N, be the i-th \<column reference\>.
>    GROUPING ( CR1, ..., CRN-1, CRN )
> is equivalent to:
>    CAST ( ( 2 * GROUPING ( CR1, ..., CRN-1 ) + GROUPING ( CRN ) ) AS IDT ) where IDT is the implementation-defined declared type of the result.